addons/website_sale/models/sale_order.py

```diff
@@ -29,6 +29,19 @@ class SaleOrder(models.Model):
     website_id = fields.Many2one('website', string='Website', readonly=True,
                                  help='Website through which this order was placed.')
 
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            if vals.get('website_id'):
+                website = self.env['website'].browse(vals['website_id'])
+                if 'company_id' in vals:
+                    company = self.env['res.company'].browse(vals['company_id'])
+                    if website.company_id.id != company.id:
+                        raise ValueError(_("The company of the website you are trying to sale from (%s) is different than the one you want to use (%s)") % (website.company_id.name, company.name))
+                else:
+                    vals['company_id'] = website.company_id.id
+        return super().create(vals_list)
```

Comment:
I haven't reviewed deeply yet but I wonder why this is necessary.
Aren't eCommerce SOs guaranteed to have both `website_id` and `company_id`? (from `_prepare_sale_order_values`)
In which case does line 42 actually change something?

Reply:
@Williambraecky Thanks for the review!
> Aren't eCommerce SOs guaranteed to have both website_id and company_id? (from _prepare_sale_order_values)

Nope, unfortunately. Maybe we should also fix in website_sale so that company_id is always set when creating SO from website, I haven't check yet all flow that call `_prepare_sale_order_values` when creating SO from website.

> In which case does line 42 actually change something?

If we set a website_id when creating an SO without setting the company_id, it will by default get the env company
https://github.com/odoo/odoo/blob/2e351d6ff9b7f72b486e83c827ff6e15a05c3331/addons/sale/models/sale_order.py#L263

Reply:
Alright, the probability is low but still I think it's ok to keep then.